includes/Google_Categories.php

```diff
@@ -0,0 +1,333 @@
+
+	/** @var string the WordPress option name where the last time the full categories list get updated is stored */
+	const OPTION_GOOGLE_PRODUCT_CATEGORIES_UPDATED = 'wc_facebook_google_product_categories_update';
+
+	/** @var array 2nd level of cache for loading Google categories list */
+	private $categories_list;
+
+
+	/**
+	 * Gets the categories list.
+	 *
+	 * @since 2.2.1-dev.1
+	 *
+	 * @return array
+	 */
+	public function get_categories() {
+
+		if ( ! empty( $this->categories_list ) ) {
+			return $this->categories_list;
+		}
+
+		// only fetch again if not fetched less than one week ago
+		$last_updated = get_transient( self::OPTION_GOOGLE_PRODUCT_CATEGORIES_UPDATED );
+
+		if ( empty ( $last_updated ) ) {
+
+			// fetch from the URL
+			$categories = $this->fetch_categories_list_from_url();
+
+			if ( ! empty( $categories ) && ! is_wp_error( $categories ) ) {
+
+				// store into database for later use
+				$this->store_categories_list( $categories );
+
+				// mark when it's stored
+				set_transient( self::OPTION_GOOGLE_PRODUCT_CATEGORIES_UPDATED, current_time( 'mysql' ), WEEK_IN_SECONDS );
+			}
+		} else {
+
+			// load from database/cached
+			$categories = $this->get_cached_categories_list();
+
+		}
+
+		$this->categories_list = $categories;
+
+		return $categories;
+	}
+
+
+	/**
+	 * Gets cached categories list form database.
+	 *
+	 * @since 2.2.1-dev.1
+	 *
+	 * @return array
+	 */
+	private function get_cached_categories_list() {
+
+		$raw_categories_items = $this->get_items();
+
+		if ( empty( $raw_categories_items ) ) {
+			return [];
+		}
+
+		$categories = [];
+
+		// structure all categories semi-flat
+		foreach ( $raw_categories_items as $category_item ) {
+			$categories[ (int) $category_item['id'] ] = [
+				'label'   => $category_item['label'],
+				'options' => [],
+				'parent'  => (int) $category_item['parent_id'],
+			];
+		}
+
+		// then insert sub-categories as options
+		foreach ( $raw_categories_items as $category_item ) {
+
+			$parent_category_id = (int) $category_item['parent_id'];
+
+			if ( isset( $categories[ $parent_category_id ] ) ) {
+				$categories[ $parent_category_id ]['options'][ (int) $category_item['id'] ] = $category_item['label'];
+			}
+		}
+
+		return $categories;
+	}
+
+
+	/**
+	 * Queries full categories list from database.
+	 *
+	 * @since 2.2.1-dev.1
+	 *
+	 * @return array
+	 */
+	private function get_items() {
+
+		global $wpdb;
+
+		return (array) $wpdb->get_results( 'SELECT * FROM ' . self::get_table_name() . ' ORDER BY id ASC', ARRAY_A );
+	}
+
+
+	/**
+	 * Fetches the full categories list from Google.
+	 *
+	 * @since 2.2.1-dev.1
+	 *
+	 * @param array $categories fetched Google categories list to store
+	 */
+	private function store_categories_list( $categories ) {
+
+		global $wpdb;
+
+		$this->empty_db_table();
+
+		foreach ( $categories as $category_id => $category ) {
+
+			$this->store_item( $category_id, $category['label'], $category['parent'] );
```

Comment:
@unfulvio I think because it's a private method call internally and it's highly unlikely that happens but I agree I should check for the keys => af6ae9a